hail/src/main/scala/is/hail/expr/ir/MatrixIR.scala

```diff
@@ -615,9 +615,17 @@ case class MatrixRepartition(child: MatrixIR, n: Int, strategy: Int) extends Mat
 
 case class MatrixUnionRows(children: IndexedSeq[MatrixIR]) extends MatrixIR {
   require(children.length > 1)
-  require(children.map(_.typ).toSet.size == 1, children.map(_.typ))
+  require(canUnionTypes, children.map(_.typ))
   val typ: MatrixType = children.head.typ
 
+  def canUnionTypes: Boolean = {
+    val types = children.map(_.typ)
+    types.map(_.colKeyStruct).toSet.size == 1 &&
+      types.map(_.rowType).toSet.size == 1 &&
+      types.map(_.rowKey).toSet.size == 1 &&
+      types.map(_.entryType).toSet.size == 1
```

Comment:
This seems inefficient. It probably doesn't matter most of the time, but I think there are uses where `children` is quite large. What about
```scala
assert(children.tail.forall(child => compatible(child.typ, children.head.typ))
def compatible(t1: MatrixType, t2: MatrixType): Boolean = {
  t1.colKeyStruct == t2.colKeyStruct &&
    ...
}
```

Reply:
👍